simulate/sam.go

```diff
@@ -13,21 +13,50 @@ import (
 	"github.com/vertgenlab/gonomics/sam"
 )
 
-// IlluminaPairedSam generates a pair of sam reads randomly distributed across the input ref.
+// IlluminaPairedSam generates pairs of sam reads randomly distributed across the input DNA sequence.
+// The inputs are the name of the input DNA sequence, the sequence itself, the number of read pairs to generate, the length of each read,
+// the average fragment size, the standard deviation of fragment sizes, the error rate where a base in
+// the read will not match the input DNA, a numbers.binomialAlias that is used to speed up calculations, and
+// output file handles for sam, bam, and a bool denoting if bam (or sam) should be the output.
+// Whichever handle (sam or bam) is not being used can be nil.
 func IlluminaPairedSam(refName string, ref []dna.Base, numPairs, readLen, avgFragmentSize int, avgFragmentStdDev float64, flatErrorRate float64, ancientErrorRate float64, flatBinomialAlias numbers.BinomialAlias, ancientBinomialAlias numbers.BinomialAlias, geometricParam float64, out *fileio.EasyWriter, bw *sam.BamWriter, bamOutput bool, deaminationDistributionSlice []int) {
-	var fragmentSize, midpoint, startFor, startRev, endFor, endRev int
+	var fragmentSize, midpoint, startFor, endRev, newCapacity int
 	var currFor, currRev sam.Sam
+	var fragment []dna.Base = make([]dna.Base, 0, avgFragmentSize+int(5*avgFragmentStdDev)) //set initial fragment slice capacity to 5 deviations above average fragment size
+	var newFragment []dna.Base                                                              //this will be used if we run out of capacity in 'fragment'
+	if avgFragmentSize < readLen {
+		log.Fatalf("Error: average fragment size %v is less than read length %v.\n", avgFragmentSize, readLen)
+	}
 
 	for i := 0; i < numPairs; i++ {
 		fragmentSize = numbers.Max(readLen, int(numbers.SampleInverseNormal(float64(avgFragmentSize), avgFragmentStdDev)))
 		midpoint = numbers.RandIntInRange(0, len(ref))
-		startFor = midpoint - (fragmentSize / 2)
-		endFor = startFor + readLen
-		endRev = midpoint + (fragmentSize / 2)
-		startRev = endRev - readLen
+		startFor = numbers.Max(midpoint-(fragmentSize/2), 0)
+		endRev = numbers.Min(midpoint+(fragmentSize/2), len(ref))
+
+		if fragmentSize < readLen {
+			readLen = fragmentSize
+		}
+
+		//here we check if fragment has sufficient capacity for the next fragment we drew. If not, we extend capacity.
+		if len(fragment)+fragmentSize > cap(fragment) {
+			newCapacity = len(fragment) + fragmentSize
+			newFragment = make([]dna.Base, fragmentSize, newCapacity)
+			fragment = newFragment
+		} else {
+			fragment = fragment[:fragmentSize]
+		}
+		copy(fragment, ref[startFor:endRev])
```

Comment:
It looks like the fragmentSize can be smaller than readLen, in which case readLen becomes fragmentSize (but average fragmentSize can't be smaller than readLen). Why is this?

What do "fragment capacity, fragment length, and fragment size" mean?